plugins/fluentd_telemetry_plugin/src/streamer.py

```diff
@@ -702,44 +865,59 @@ def stream_data(self, telemetry_endpoint):  # pylint: disable=too-many-locals
+                    self._parse_telemetry_prometheus_metrics_to_json(telemetry_data, msg_tag) \
                     if ufm_telemetry_is_prometheus_format else \
-                    self._parse_telemetry_csv_metrics_to_json(telemetry_data)
+                    self._parse_telemetry_csv_metrics_to_json(telemetry_data, msg_tag, is_xdr_mode)
                 end_time = time.time()
                 data_len = len(data_to_stream)
                 resp_process_time = round(end_time - start_time, 6)
                 self.streaming_metrics_mgr.update_streaming_metrics(msg_tag, **{
                     self.streaming_metrics_mgr.telemetry_response_process_time_seconds_key: resp_process_time
                 })
-                if data_len > 0 and \
-                        (not new_data_timestamp or
-                         (new_data_timestamp and new_data_timestamp != self.last_streamed_data_sample_timestamp)):
-                    self.streaming_metrics_mgr.update_streaming_metrics(msg_tag, **{
-                        self.streaming_metrics_mgr.num_of_streamed_ports_in_last_msg_key: data_len,
-                        self.streaming_metrics_mgr.num_of_processed_counters_in_last_msg_key: num_of_counters
-                    })
-                    logging.info('Processing of endpoint %s Completed In: %.2f Seconds. '
-                                 '(%d) Ports, (%d) Counters Were Handled',
-                                 msg_tag, resp_process_time, data_len, num_of_counters)
-                    if self.bulk_streaming_flag:
-                        self._stream_data_to_fluentd(data_to_stream, msg_tag)
-                    else:
-                        for row in data_to_stream:
-                            self._stream_data_to_fluentd(row, msg_tag)
-                    self.last_streamed_data_sample_timestamp = new_data_timestamp
-                elif self.stream_only_new_samples:
-                    logging.info('No new samples in endpoint %s, nothing to stream', msg_tag)
-
-            except Exception as ex:  # pylint: disable=broad-except
-                logging.error("Exception occurred during parsing telemetry data: %s", str(ex))
-        else:
-            logging.error("Failed to get the telemetry data metrics for %s", _url)
+                logging.info('Processing of endpoint %s Completed In: %.2f Seconds. '
+                             '(%d) Ports, (%d) Counters Were Handled',
+                             msg_tag, resp_process_time, data_len, num_of_counters)
+            elif self.stream_cached_data_on_telemetry_fail_is_enabled and \
+                    self.last_streamed_data_sample_per_endpoint.get(msg_tag):
+                # if the telemetry endpoint currently unavailable, try to get the cached data
+                data_to_stream = list(self.last_streamed_data_sample_per_endpoint[msg_tag].values())
+                data_len = len(data_to_stream)
+                if data_len > 0:
```

Comment:
why the condition is data_len > 0 and no data_len == 0 for unavailable endpoint?

Reply:
data_len > 0 seems to be check for last valid cached data that should be streamed if there is no telemetry.

Reply:
As @vg12345 mentioned, this check is applied to verify if there's cached data for the specific endpoint or not